File: src/SourceLanguageCatalog.pas
unit SourceLanguageCatalog;

{$IFDEF FPC}
{$MODE OBJFPC}{$H+}
{$CODEPAGE UTF8}
{$ENDIF}

interface

uses
  SysUtils;

const
  SOURCE_LANGUAGE_CATALOG_VERSION = 'pt-PT-1.0.0';
  CUTLET_NAME_COUNT = 17;
  MONTH_NAME_COUNT = 47;

function CutletNameByCanonicalIndex(Index: Integer): UnicodeString;
function MonthNameByCanonicalIndex(Index: Integer): UnicodeString;
function SourceLanguageCatalogIsFrozen: Boolean;
function SourceLanguageCatalogSelfCheck: Boolean;

implementation

var
  CutletNames: array[1..CUTLET_NAME_COUNT] of UnicodeString;
  MonthNames: array[1..MONTH_NAME_COUNT] of UnicodeString;
  Frozen: Boolean = False;

procedure InitializeCatalog;
begin
  { O índice canónico é a única ordem normativa. O texto nunca participa em rank, unrank ou chaves semânticas. }
  CutletNames[1] := 'bronze';
  CutletNames[2] := 'raposa';
  CutletNames[3] := 'rim';
  CutletNames[4] := 'lárix';
  CutletNames[5] := 'pensamento';
  CutletNames[6] := 'quatro partes de nove';
  CutletNames[7] := 'Palgurash';
  CutletNames[8] := 'junco';
  CutletNames[9] := 'cacho';
  CutletNames[10] := 'escorpião';
  CutletNames[11] := 'cinza';
  CutletNames[12] := 'trigo';
  CutletNames[13] := 'rio';
  CutletNames[14] := 'riso';
  CutletNames[15] := 'Acade';
  CutletNames[16] := 'chifre';
  CutletNames[17] := 'o jarro vazio';

  MonthNames[1] := 'barro';
  MonthNames[2] := 'romã';
  MonthNames[3] := 'cotovelo';
  MonthNames[4] := 'inveja';
  MonthNames[5] := 'Eridu';
  MonthNames[6] := 'pasta de dentes';
  MonthNames[7] := 'três partes de cinco';
  MonthNames[8] := 'Karshumav';
  MonthNames[9] := 'leopardo';
  MonthNames[10] := 'estanho';
  MonthNames[11] := 'nevoeiro';
  MonthNames[12] := 'olíbano';
  MonthNames[13] := 'fuso';
  MonthNames[14] := 'costela';
  MonthNames[15] := 'alfarroba';
  MonthNames[16] := 'Uruk';
  MonthNames[17] := 'vergonha';
  MonthNames[18] := 'camelo';
  MonthNames[19] := 'cobre';
  MonthNames[20] := 'poço';
  MonthNames[21] := 'gema';
  MonthNames[22] := 'estrela';
  MonthNames[23] := 'mel';
  MonthNames[24] := 'baço';
  MonthNames[25] := 'calcário';
  MonthNames[26] := 'alegria';
  MonthNames[27] := 'figo';
  MonthNames[28] := 'Nínive';
  MonthNames[29] := 'rã';
  MonthNames[30] := 'piche';
  MonthNames[31] := 'vela';
  MonthNames[32] := 'a porta fechada';
  MonthNames[33] := 'gergelim';
  MonthNames[34] := 'nuca';
  MonthNames[35] := 'prata';
  MonthNames[36] := 'lírio';
  MonthNames[37] := 'tempestade';
  MonthNames[38] := 'burro';
  MonthNames[39] := 'farinha';
  MonthNames[40] := 'arrependimento';
  MonthNames[41] := 'Babilónia';
  MonthNames[42] := 'língua';
  MonthNames[43] := 'linho';
  MonthNames[44] := 'sal';
  MonthNames[45] := 'pera';
  MonthNames[46] := 'arco';
  MonthNames[47] := 'areia';

  Frozen := True;
end;

function CutletNameByCanonicalIndex(Index: Integer): UnicodeString;
begin
  if (Index < 1) or (Index > CUTLET_NAME_COUNT) then
    raise ERangeError.Create('Índice canónico de costeleta fora do intervalo.');
  Result := CutletNames[Index];
end;

function MonthNameByCanonicalIndex(Index: Integer): UnicodeString;
begin
  if (Index < 1) or (Index > MONTH_NAME_COUNT) then
    raise ERangeError.Create('Índice canónico de mês fora do intervalo.');
  Result := MonthNames[Index];
end;

function SourceLanguageCatalogIsFrozen: Boolean;
begin
  Result := Frozen;
end;

function SourceLanguageCatalogSelfCheck: Boolean;
var
  I, J: Integer;
begin
  Result := Frozen;
  if not Result then Exit;
  for I := 1 to CUTLET_NAME_COUNT do
  begin
    if CutletNames[I] = '' then Exit(False);
    for J := I + 1 to CUTLET_NAME_COUNT do
      if CutletNames[I] = CutletNames[J] then Exit(False);
  end;
  for I := 1 to MONTH_NAME_COUNT do
  begin
    if MonthNames[I] = '' then Exit(False);
    for J := I + 1 to MONTH_NAME_COUNT do
      if MonthNames[I] = MonthNames[J] then Exit(False);
  end;
  Result := True;
end;

initialization
  InitializeCatalog;

end.

File: src/source_language_catalog.pl
:- module(source_language_catalog,
    [ source_language_catalog_version/1,
      cutlet_source_name/2,
      month_source_name/2,
      cutlet_source_names/1,
      month_source_names/1,
      canonical_cutlet_indices/1,
      canonical_month_indices/1
    ]).

source_language_catalog_version('1.0.0').

cutlet_source_name(1,  'bronce').
cutlet_source_name(2,  'raposo').
cutlet_source_name(3,  'ril').
cutlet_source_name(4,  'Lagash').
cutlet_source_name(5,  'pensamento').
cutlet_source_name(6,  'catro partes de nove').
cutlet_source_name(7,  'Palgurash').
cutlet_source_name(8,  'xunco').
cutlet_source_name(9,  'acio').
cutlet_source_name(10, 'escorpión').
cutlet_source_name(11, 'cinza').
cutlet_source_name(12, 'trigo').
cutlet_source_name(13, 'río').
cutlet_source_name(14, 'risa').
cutlet_source_name(15, 'Acad').
cutlet_source_name(16, 'corno').
cutlet_source_name(17, 'o cántaro baleiro').

month_source_name(1,  'barro').
month_source_name(2,  'granada').
month_source_name(3,  'cóbado').
month_source_name(4,  'envexa').
month_source_name(5,  'Eridu').
month_source_name(6,  'pasta de dentes').
month_source_name(7,  'tres partes de cinco').
month_source_name(8,  'Karshumab').
month_source_name(9,  'tigre').
month_source_name(10, 'estaño').
month_source_name(11, 'néboa').
month_source_name(12, 'incenso').
month_source_name(13, 'fuso').
month_source_name(14, 'costela').
month_source_name(15, 'alfarroba').
month_source_name(16, 'Uruk').
month_source_name(17, 'vergoña').
month_source_name(18, 'camelo').
month_source_name(19, 'cobre').
month_source_name(20, 'pozo').
month_source_name(21, 'xema').
month_source_name(22, 'estrela').
month_source_name(23, 'mel').
month_source_name(24, 'bazo').
month_source_name(25, 'pedra calcaria').
month_source_name(26, 'alegría').
month_source_name(27, 'figo').
month_source_name(28, 'Nínive').
month_source_name(29, 'ra').
month_source_name(30, 'alcatrán').
month_source_name(31, 'vela').
month_source_name(32, 'a porta pechada').
month_source_name(33, 'sésamo').
month_source_name(34, 'caluga').
month_source_name(35, 'prata').
month_source_name(36, 'lirio').
month_source_name(37, 'treboada').
month_source_name(38, 'burro').
month_source_name(39, 'fariña').
month_source_name(40, 'arrepentimento').
month_source_name(41, 'Babilonia').
month_source_name(42, 'lingua').
month_source_name(43, 'liño').
month_source_name(44, 'sal').
month_source_name(45, 'pera').
month_source_name(46, 'arco').
month_source_name(47, 'area').

canonical_cutlet_indices([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17]).
canonical_month_indices([
    1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,
    25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47
]).

cutlet_source_names(Names) :-
    canonical_cutlet_indices(Indices),
    names_from_indices(cutlet, Indices, Names).

month_source_names(Names) :-
    canonical_month_indices(Indices),
    names_from_indices(month, Indices, Names).

names_from_indices(_, [], []).
names_from_indices(cutlet, [I|Is], [Name|Names]) :-
    cutlet_source_name(I, Name),
    names_from_indices(cutlet, Is, Names).
names_from_indices(month, [I|Is], [Name|Names]) :-
    month_source_name(I, Name),
    names_from_indices(month, Is, Names).
